### PR TITLE
Add the MonoTime equivalents of std.datetime.StopWatch/benchmark.

### PR DESCRIPTION
**Pull request by jmdavis**
std.datetime.package has StopWatch, benchmark, comparingBenchmark, and
measureTime, all of which use TickDuration (which would be deprecated,
but it can't be deprecated as long as those functions in std.datetime
are deprecated). This commit introduces
std.datetime.stopwatch to replace those functions in std.datetime. In
order to avoid symbol conflicts, std.datetime.stopwatch will not be
publicly import in std.datetime.package until the old symbols have been
removed.

std.datetime.experimental.stopwatch contains StopWatch and benchmark
which have essentially the same APIs as the ones in
std.datetime.package, but they use MonoTime and Duration.

comparingBenchmark has not been ported to MonoTime and Duration, because
it is simply a wrapper around benchmark.

measureTime has not been ported to MonoTime and Duration, because it is
equivalent to using StopWatch with a scope(exit) statement.

The old functionality will be deprecated the major release after the new
symbols have been introduced.

### CODE REVIEW

**Comment by jmdavis** [comment]
FYI: I've already talked about these changes with Andrei, and he's okay with them.

File: std/datetime/stopwatch.d [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+Duration[fun.length] benchmark(fun...)(uint n)
+{
+    Duration[fun.length] result;
+    auto sw = StopWatch(AutoStart.yes);
+
+    foreach (i, unused; fun)
+    {
+        sw.reset();
+        foreach (_; 0 .. n)
+            fun[i]();
```
Comment (wilzbach) [inline std/datetime/stopwatch.d]: One minor thing that one nearly always has to think about is that a reasonably smart optimizer will just remove this code block.
There are a couple of tricks to avoid this behavior, at least:
- assign return value to `__gshared`
- use the `doNotOptimizeAway` trick (see below or #2995)
- [`@weak` in LDC](https://wiki.dlang.org/LDC-specific_language_changes#.40.28ldc.attributes.weak.29)
- ...

Imho it would make `benchmark` a lot easier to use if this would be done automatically.
FWIW Andrei [mentioned once](https://github.com/dlang/phobos/pull/2995#issuecomment-152248649):

> but doesn't include things like "doNotOptimizeAway" which is essential.

```d
private void doNotOptimizeAway(T)(auto ref T t)
{
    import core.thread : getpid;
    import std.stdio : writeln;
    if(getpid() == 1) {
        writeln(*cast(char*)&t);
    }
```

File: std/datetime/stopwatch.d [hunk 2]
```diff
@@ -0,0 +1,425 @@
+// Written in the D programming language
+
+/++
+    Module containing some basic benchmarking and timing functionality.
+
+    For convenience, this module publicly imports core.time.
```
Comment (wilzbach) [inline std/datetime/stopwatch.d]: Nit: For consistency core.time could be a `MREF`...

File: std/datetime/stopwatch.d [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+
+    /++
+       Starts the StopWatch.
+
+       start should not be called if the StopWatch is already running.
+      +/
+    void start() @safe nothrow @nogc
+    in { assert(!running); }
```
Comment (wilzbach) [inline std/datetime/stopwatch.d]: There was a longer discussion this summer on which a majority agreed that an `assert` statement should have a message even if it's trivial as it greatly helps the user in figuring out what went wrong instead of the bare segfault.

File: std/datetime/stopwatch.d [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+
+    /++
+       Stops the StopWatch.
+
+       stop should not be called if the StopWatch is not running.
+      +/
+    void stop() @safe nothrow @nogc
+    in { assert(_running); }
```
Comment (wilzbach) [inline std/datetime/stopwatch.d]: ditto about `assert` with a error message

File: std/datetime/stopwatch.d [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+
+    foreach (i, unused; fun)
+    {
+        sw.reset();
+        foreach (_; 0 .. n)
+            fun[i]();
+        result[i] = sw.peek();
+    }
+
+    return result;
```
Comment (wilzbach) [inline std/datetime/stopwatch.d]: May I ask what the rationale for removing `opEquals` was?

File: std/datetime/stopwatch.d [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+    }
+
+    /++
+        Constructs a StopWatch. Whether it starts immediately depends on the
+        $(LREF AutoStart) argument.
+
+        If $(D StopWatch.init) is used, then the constructed StopWatch isn't
+        running (and can't be, since no constructor ran).
+      +/
+    this(AutoStart autostart) @safe nothrow @nogc
```
Comment (jmdavis) [inline std/datetime/stopwatch.d]: Aside from whether that would help or harm usability, it wouldn't work in D, because then it would be a default constructor.

File: std/datetime/stopwatch.d (hunk 1 above)
Comment (jmdavis) [inline std/datetime/stopwatch.d]: I'd prefer to leave that improvement to a separate PR. The focus of this is to give us what we already have but using `MonoTime` and `Duration` rather than `TickDuration`.

File: std/datetime/stopwatch.d (hunk 5 above)
Comment (jmdavis) [inline std/datetime/stopwatch.d]: I probably removed it with the idea that it did the same thing as the defaults, since when `StopWatch` was originally added, it was not uncommon that you had to declare it yourself. Looking over it now, it looks like the original implementation is different from the default in that it skips comparing the `_running` member, therefore not treating whether the `StopWatch` is running as part of its comparable state. I don't know if that makes sense or not. I strongly question whether it makes any sense to compare a `StopWatch` in the first place.

**Comment by jmdavis** [comment]
Updated based on Sebastian's comments.
